### PR TITLE
Optional segment load/drop management without zookeeper using http

### PR DESCRIPTION
**Pull request by himanshug**
~~TODOs:~~
- ~~Test on a cluster with decent load.~~

This patch introduces following changes.

On Historical Side:
1) All non-zookeeper related code is extracted out of ZkCoordinator and moved into new class SegmentLoadDropHandler. ZkCoordinator now uses SegmentLoadDropHandler to delegate load/drop requests.
2) A new method `ListenableFuture SegmentLoadDropHandler.processBatch(List<DataSegmentChangeRequest>)` is added to support http endpoint for segment load/drop management.
3) New endpoint `POST /druid-internal/v1/segments/changeRequests` is added in SegmentListerResource that uses `ListenableFuture SegmentLoadDropHandler.processBatch(List<DataSegmentChangeRequest>)` to delegate batch of load/drop requests received. Here is a copy-paste of javadoc for this endpoint.
``` 
  /**
   * This endpoint is used by HttpLoadQueuePeon to assign segment load/drop requests batch. This endpoint makes the
   * client wait till one of the following events occur. Note that this is implemented using async IO so no jetty
   * threads are held while in wait.
   *
   * (1) Given timeout elapses.
   * (2) Some load/drop request completed.
   *
   * It returns a map of "load/drop request -> SUCCESS/FAILED/PENDING status" for each request in the batch.
   */
```
4) `druid.segmentCache.numLoadingThreads` configuration is revived and must be greater than the `"druid.coordinator.loadqueuepeon.http.batchSize` described later.

On Coordinator Side:
1) CuratorLoadQueuePeon is created by copying existing LoadQueuePeon class.
2) LoadQueuePeon is made abstract class.
3) HttpLoadQueuePeon is created to do segment load/drop management via http. It uses the new endpoint introduced above to send batch of load/drop requests to Historical node.
4) Following new configurations are introduced.
`druid.coordinator.loadqueuepeon.type` = http or curator, curator by default
`druid.coordinator.loadqueuepeon.http.batchSize` = number of load/drop to try and process in parallel, must be less than `druid.segmentCache.numLoadingThreads` on historical
`druid.coordinator.loadqueuepeon.http.repeatDelay` = delay Duration for periodic check for new load/drop request to send to historical. note this could be very large given that same code is executed whenever a new load/drop is requested to HttpLoadQueuePeon without waiting for schedule to kick in, default 1 minute
`druid.coordinator.loadqueuepeon.http.hostTimeout` = default 5 minutes, timeout used to be specified on new endpoint introduced for batch load/drop request.

In the long run, I plan to remove CuratorLoadQueuePeon and ZkCoordinator. They are being kept now only to be backward compatible and for HttpLoadQueuePeon to prove itself in production.

Documentation for HttpLoadQueuePeon is intentionally left out.

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+  {
+    synchronized (lock) {
+      if (stopped) {
+        return;
+      }
+
+      stopped = true;
+
+      for (SegmentHolder holder : segmentsToDrop.values()) {
+        holder.requestCompleted(false, null);
```
Comment (cheddar) [inline server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: This is acting like these requests succeeded, is that the correct thing to do?

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+  }
+
+  @Override
+  public int getAndResetFailedAssignCount()
+  {
+    return failedAssignCount.getAndSet(0);
+  }
+
+  @Override
+  public void markSegmentToDrop(DataSegment dataSegment)
```
Comment (cheddar) [inline server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: What's the difference between this and dropSegment?

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+  }
+
+  @Override
+  public int getNumberOfSegmentsInQueue()
+  {
+    return segmentsToLoad.size();
+  }
+
+  @Override
+  public Set<DataSegment> getSegmentsMarkedToDrop()
```
Comment (cheddar) [inline server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: Should this really be returning the set, which is likely mutable?

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+  private class SegmentHolder
+  {
+    private final DataSegment segment;
+    private final DataSegmentChangeRequest changeRequest;
+    private final int type;
+    private final List<LoadPeonCallback> callbacks = Lists.newArrayList();
+
+    private SegmentHolder(
+        DataSegment segment,
+        int type,
```
Comment (cheddar) [inline server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: What's the point of this type thing?  We should either leverage polymorphism or completely eliminate it.  Using it sometimes and not other times gets confusing

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+              return curr;
+            }
+          }
+      );
+    }
+
+    public void requestCompleted(boolean failed, String failMsg)
+    {
+      synchronized (lock) {
+        Future cancellationFuture = cancellationFutureRef.get();
```
Comment (cheddar) [inline server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: I really don't understand this cancellation future stuff....

File: server/src/main/java/io/druid/segment/loading/SegmentLoaderConfig.java [hunk 6]
```diff
@@ -46,12 +44,18 @@
...
+  private int numLoadingThreads = 1;
+
   @JsonProperty("numBootstrapThreads")
   private Integer numBootstrapThreads = null;
 
   @JsonProperty
   private File infoDir = null;
 
+  @JsonProperty
+  private int statusQueueMaxSize = 100;
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/segment/loading/SegmentLoaderConfig.java]: yes, will let it run on our servers for a while before documenting. hopefully we'll have enough confidence in this before 0.11.1 release so will get documented before that release.

File: java-util/src/main/java/io/druid/java/util/common/Pair.java [hunk 7]
```diff
@@ -33,10 +34,16 @@
...
   }
 
+  @JsonProperty
   public final T1 lhs;
+
+  @JsonProperty
   public final T2 rhs;
 
-  public Pair(T1 lhs, T2 rhs)
+  public Pair(
```
[the remark this answers is not in the document]
Comment (himanshug) [reply java-util/src/main/java/io/druid/java/util/common/Pair.java]: not using Pair anymore.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 8]
```diff
@@ -0,0 +1,797 @@
...
+
+  @LifecycleStart
+  public void start() throws IOException
+  {
+    synchronized (lock) {
+      if (started) {
+        return;
+      }
+
+      log.info("Starting...");
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: no, just wanted to mark start/stop in log to check whether it indeed gets started/stopped etc.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 9]
```diff
@@ -0,0 +1,797 @@
...
+      log.info("Stopped.");
+    }
+  }
+
+  public boolean isStarted()
+  {
+    return started;
+  }
+
+  private void loadLocalCache()
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: unrelated to work in this PR. its a copy paste from ZkCoordinator and this PR is keeping old things exactly same.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 10]
```diff
@@ -0,0 +1,797 @@
...
+          removeSegment(segment, callback, false);
+          throw new SegmentLoadingException(
+              e, "Failed to write to disk segment info cache file[%s]", segmentInfoCacheFile
+          );
+        }
+      }
+    }
+  }
+
+  void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: not anymore ... I made SegmentLoadDropHandler to implement DataSegmentChangeHandler as was done in old ZkCoordinator. this is public now and is tested via the UTs in SegmentLoadDropHandlerTest which are copied from earlier version of ZkCoordinatorTest.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 11]
```diff
@@ -0,0 +1,797 @@
...
+         and if (segmentsToDelete.remove(segment)) returns true, in which case historical will start deleting segment
+         files. At that point, it's possible that right after the "if" check, addSegment() is called and actually loads
+         the segment, which makes dropping segment and downloading segment happen at the same time.
+       */
+      if (segmentsToDelete.contains(segment)) {
+        /*
+           Both contains(segment) and remove(segment) can be moved inside the synchronized block. However, in that case,
+           each time when addSegment() is called, it has to wait for the lock in order to make progress, which will make
+           things slow. Given that in most cases segmentsToDelete.contains(segment) returns false, it will save a lot of
+           cost of acquiring lock by doing the "contains" check outside the synchronized block.
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: maybe but unrelated to work in this PR. its a copy paste from ZkCoordinator and this PR is keeping old things exactly same.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 12]
```diff
@@ -0,0 +1,797 @@
...
+    }
+    finally {
+      callback.execute();
+      if (loadingExecutor != null) {
+        loadingExecutor.shutdownNow();
+      }
+    }
+  }
+
+  void removeSegment(
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: not anymore ... I made SegmentLoadDropHandler to implement DataSegmentChangeHandler as was done in old ZkCoordinator. this is public now and is tested via the UTs in SegmentLoadDropHandlerTest which are copied from earlier version of ZkCoordinatorTest.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 13]
```diff
@@ -0,0 +1,797 @@
...
+               .emit();
+            result = Status.failed(e.getMessage());
+          }
+          finally {
+            updateStatusAndResolveWaitingFutures(new SegmentChangeRequestDrop(segment), result);
+          }
+        }
+      };
+
+      if (scheduleDrop) {
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: fixed

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 14]
```diff
@@ -0,0 +1,797 @@
...
+  public Collection<DataSegment> getPendingDeleteSnapshot()
+  {
+    return ImmutableList.copyOf(segmentsToDelete);
+  }
+
+  public ListenableFuture<List<Pair<DataSegmentChangeRequest, Status>>> processBatch(List<DataSegmentChangeRequest> changeRequests)
+  {
+    boolean isAnyRequestDone = false;
+
+    Map<DataSegmentChangeRequest, AtomicReference<Status>> statuses = new HashMap<>(changeRequests.size());
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: changed.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 15]
```diff
@@ -0,0 +1,797 @@
...
+    }
+
+    return future;
+  }
+
+  private AtomicReference<Status> processRequest(DataSegmentChangeRequest changeRequest)
+  {
+    synchronized (requestStatusesLock) {
+      if (requestStatuses.getIfPresent(changeRequest) == null) {
+        if (changeRequest instanceof SegmentChangeRequestLoad) {
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: changed

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 16]
```diff
@@ -0,0 +1,797 @@
...
+          requestStatuses.put(changeRequest, new AtomicReference<>(Status.PENDING));
+          exec.submit(
+              () -> SegmentLoadDropHandler.this.addSegment(
+                  ((SegmentChangeRequestLoad) changeRequest).getSegment(),
+                  DataSegmentChangeCallback.NOOP
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: using callback to resolve futures now.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 17]
```diff
@@ -0,0 +1,797 @@
...
+          );
+        } else {
+          throw new IAE("Not a segment load/drop request[%s].", changeRequest.getClass().getName());
+        }
+      }
+      return requestStatuses.getIfPresent(changeRequest);
+    }
+  }
+
+  private void updateStatusAndResolveWaitingFutures(DataSegmentChangeRequest chanteRequest, Status result)
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: fixed

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 18]
```diff
@@ -0,0 +1,603 @@
...
+    this.processingExecutor = processingExecutor;
+    this.callBackExecutor = callBackExecutor;
+
+    this.serverId = baseUrl;
+    try {
+      this.changeRequestURL = new URL(
+          new URL(baseUrl),
+          StringUtils.safeFormat(
+              "druid-internal/v1/segments/changeRequests?timeout=%d",
+              config.getHttpLoadQueuePeonHostTimeout().getMillis()
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: no

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 19]
```diff
@@ -0,0 +1,603 @@
...
+      log.debug("[%s]Ignoring tick. Either in-progress already or stopped.", serverId);
+      return;
+    }
+
+    int batchSize = config.getHttpLoadQueuePeonBatchSize();
+
+    List<DataSegmentChangeRequest> newRequests = new ArrayList<>(batchSize);
+
+    synchronized (lock) {
+      Iterator<Map.Entry<DataSegment, SegmentHolder>> iter = segmentsToDrop.entrySet().iterator();
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: changed

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 20]
```diff
@@ -0,0 +1,603 @@
...
+      while (batchSize-- > 0 && iter.hasNext()) {
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: changed

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 21]
```diff
@@ -0,0 +1,603 @@
...
+
+    if (newRequests.size() > 0) {
+      try {
+        log.debug("Sending [%d] load/drop requests to Server[%s].", newRequests.size(), serverId);
+        BytesAccumulatingResponseHandler responseHandler = new BytesAccumulatingResponseHandler();
+        ListenableFuture<InputStream> future = httpClient.go(
+            new Request(HttpMethod.POST, changeRequestURL)
+                .addHeader(HttpHeaders.Names.ACCEPT, MediaType.APPLICATION_JSON)
+                .addHeader(HttpHeaders.Names.CONTENT_TYPE, MediaType.APPLICATION_JSON)
+                .setContent(jsonMapper.writerWithType(REQUEST_ENTITY_TYPE_REF).writeValueAsBytes(newRequests)),
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: changed

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 22]
```diff
@@ -0,0 +1,603 @@
...
+                      if (stopped) {
+                        return;
+                      }
+
+                      for (Pair<DataSegmentChangeRequest, SegmentLoadDropHandler.Status> e : statuses) {
+                        switch (e.rhs.getState()) {
+                          case SUCCESS:
+                          case FAILED:
+                            e.lhs.go(
+                                new DataSegmentChangeHandler()
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: done

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 23]
```diff
@@ -0,0 +1,603 @@
...
+                                    }
+                                  }
+                                }, null
+                            );
+                            break;
+                          case PENDING:
+                            log.debug("[%s]Segment request [%s] is pending.", serverId, e.lhs);
+                            break;
+                          default:
+                            scheduleNextRunImmediately = false;
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: simplified the code a bit, its should be more clear now.

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 24]
```diff
@@ -0,0 +1,603 @@
...
+                  }
+                  finally {
+                    mainLoopInProgress.set(false);
+                  }
+                } else {
+                  scheduleNextRunImmediately = false;
+                  onFailure(new RE("Unexpected Response Status."));
+                }
+
+                if (scheduleNextRunImmediately) {
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: simplified the code a bit, its should be more clear now.

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java (hunk 1 above)
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: yes, this behavior is retained from existing behavior in CuratorLoadQueuePeon (copy of older LoadQueuePeon class )

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java (hunk 2 above)
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: this behavior is retained from existing behavior in CuratorLoadQueuePeon (copy of older LoadQueuePeon class ). I am not sure about this but it was recently added so kept.

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java (hunk 3 above)
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: this behavior is retained from existing behavior in CuratorLoadQueuePeon (copy of older LoadQueuePeon class ). but changed to return unmodifiable version.

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java (hunk 4 above)
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: changed to use Polymorphism.

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java [hunk 25]
```diff
@@ -0,0 +1,603 @@
...
+                  TimeUnit.MILLISECONDS
+              );
+            } else {
+              return curr;
+            }
+          }
+      );
+    }
+
+    public void requestCompleted(boolean failed, String failMsg)
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: changed

File: server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java (hunk 5 above)
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java]: this code is removed.

File: server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java [hunk 26]
```diff
@@ -0,0 +1,797 @@
...
+  {
+    this(jsonMapper, config, announcer, serverAnnouncer, segmentManager,
+         Executors.newScheduledThreadPool(
+             config.getNumLoadingThreads(),
+             Execs.makeThreadFactory("SimpleDataSegmentChangeHandler-%s")
+         )
+    );
+  }
+
+  @VisibleForTesting
```
[the remark this answers is not in the document]
Comment (himanshug) [reply server/src/main/java/io/druid/server/coordination/SegmentLoadDropHandler.java]: This constructor requires creator to pass the `ExecutorService` which is very handy for tests to be able to deterministic. However, I don't really want any other code to use this constructor and hence added that annotation. If you still believe that its not useful then I'll remove it.

**Pull request closed without merging**